main/pipeline/pipeline.cc

```diff
@@ -428,55 +428,107 @@ struct IndexThreadResultPack {
+            Timer timeit(cgs.tracer(), "substituteTreesWorker");
+            IndexSubstitutionJob job;
+            for (auto result = batchq->try_pop(job); !result.done(); result = batchq->try_pop(job)) {
+                if (result.gotItem()) {
+                    if (job.subst.has_value()) {
+                        for (auto &tree : job.trees) {
+                            auto file = tree.file;
+                            if (!file.data(cgs).cached) {
+                                core::MutableContext ctx(cgs, core::Symbols::root(), file);
+                                tree.tree = ast::Substitute::run(ctx, *job.subst, move(tree.tree));
+                            }
                         }
                     }
+                    auto numSubstitutedTrees = job.trees.size();
+                    resultq->push(std::move(job.trees), numSubstitutedTrees);
                 }
-                ret.trees.insert(ret.trees.end(), make_move_iterator(threadResult.res.trees.begin()),
-                                 make_move_iterator(threadResult.res.trees.end()));
             }
-            progress.reportProgress(input->doneEstimate());
+        });
+
+        vector<ast::ParsedFile> trees;
+        for (auto result = resultq->wait_pop_timed(trees, WorkerPool::BLOCK_INTERVAL(), cgs.tracer()); !result.done();
+             result = resultq->wait_pop_timed(trees, WorkerPool::BLOCK_INTERVAL(), cgs.tracer())) {
+            if (result.gotItem()) {
+                ret.insert(ret.end(), std::make_move_iterator(trees.begin()), std::make_move_iterator(trees.end()));
+                progress.reportProgress(resultq->doneEstimate());
+            }
         }
     }
+
     return ret;
 }
 
-IndexResult indexSuppliedFiles(const shared_ptr<core::GlobalState> &baseGs, vector<core::FileRef> &files,
-                               const options::Options &opts, WorkerPool &workers,
-                               const unique_ptr<const OwnedKeyValueStore> &kvstore) {
-    Timer timeit(baseGs->tracer(), "indexSuppliedFiles");
+vector<ast::ParsedFile> indexSuppliedFiles(core::GlobalState &baseGs, vector<core::FileRef> &files,
+                                           const options::Options &opts, WorkerPool &workers,
+                                           const unique_ptr<const OwnedKeyValueStore> &kvstore) {
+    Timer timeit(baseGs.tracer(), "indexSuppliedFiles");
     auto resultq = make_shared<BlockingBoundedQueue<IndexThreadResultPack>>(files.size());
     auto fileq = make_shared<ConcurrentBoundedQueue<core::FileRef>>(files.size());
     for (auto &file : files) {
         fileq->push(move(file), 1);
     }
 
-    workers.multiplexJob("indexSuppliedFiles", [baseGs, &opts, fileq, resultq, &kvstore]() {
-        Timer timeit(baseGs->tracer(), "indexSuppliedFilesWorker");
-        unique_ptr<core::GlobalState> localGs = baseGs->deepCopy();
+    std::shared_ptr<core::GlobalState> emptyGs = baseGs.copyForIndex();
```

Comment:
Should we mark `emptyGs` `const`? would that give us additional guarantees that no one tries to modify it in a thread?

Reply:
I'm +1 to making GlobalStates `const` when possible.